packages/livekit-rtc/src/audio_stream.ts

```diff
@@ -47,54 +53,81 @@ export class AudioStream implements AsyncIterableIterator<AudioFrame> {
     this.info = res.stream!.info!;
     this.ffiHandle = new FfiHandle(res.stream!.handle!.id!);
 
-    FfiClient.instance.on(FfiClientEvent.FfiEvent, this.onEvent);
-  }
-
-  private onEvent = (ev: FfiEvent) => {
-    if (
-      ev.message.case != 'audioStreamEvent' ||
-      ev.message.value.streamHandle != this.ffiHandle.handle
-    ) {
-      return;
-    }
+    const infinite_capacity = capacity <= 0;
 
-    const streamEvent = ev.message.value.message;
-    switch (streamEvent.case) {
-      case 'frameReceived':
-        const frame = AudioFrame.fromOwnedInfo(streamEvent.value.frame!);
-        if (this.queueResolve) {
-          this.queueResolve({ done: false, value: frame });
-          this.queueResolve = null;
-        } else {
-          this.eventQueue.push(frame);
+    const source = new ReadableStream<FfiEvent>({
+      start: (controller) => {
+        this.onEvent = (ev: FfiEvent) => {
+          if (
+            ev.message.case === 'audioStreamEvent' &&
+            ev.message.value.streamHandle === this.ffiHandle.handle
+          ) {
+            if (infinite_capacity || (controller.desiredSize && controller.desiredSize > 0)) {
+              controller.enqueue(ev);
+            } else {
+              console.warn('Audio stream buffer is full, dropping frame');
+            }
+          }
+        };
+        FfiClient.instance.on(FfiClientEvent.FfiEvent, this.onEvent);
+      },
+      cancel: () => {
+        if (this.onEvent) {
+          FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
+          this.onEvent = null;
         }
-        break;
-      case 'eos':
-        FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
-        break;
-    }
-  };
+      },
+    });
+
+    const transformStream = new TransformStream<FfiEvent, AudioFrame>(
+      {
+        transform: (event: FfiEvent, controller: TransformStreamDefaultController<AudioFrame>) => {
+          if (
+            event.message.case !== 'audioStreamEvent' ||
+            event.message.value.streamHandle !== this.ffiHandle.handle
+          ) {
+            return;
+          }
+
+          const streamEvent = event.message.value.message;
+          switch (streamEvent.case) {
+            case 'frameReceived':
+              const frame = AudioFrame.fromOwnedInfo(streamEvent.value.frame!);
+              controller.enqueue(frame);
+              break;
+            case 'eos':
+              controller.terminate();
+              if (this.onEvent) {
+                FfiClient.instance.off(FfiClientEvent.FfiEvent, this.onEvent);
+                this.onEvent = null;
+              }
+              break;
+          }
+        },
+      },
+      {
+        highWaterMark: capacity > 0 ? capacity : undefined,
+      },
+    );
+
+    this.reader = source.pipeThrough(transformStream).getReader();
```

Comment:
we could acquire the reader lazily in `next()` and also make sure to call `releaseLock` when done